Src/Newtonsoft.Json.Schema/Infrastructure/Validation/ObjectScope.cs

```diff
@@ -28,6 +35,36 @@ public void Initialize(ContextBase context, SchemaScope? parent, int initialDept
             _propertyCount = 0;
             _currentPropertyName = null;
 
+            if (context.Validator.PropertyNameCaseInsensitive)
+            {
+                if (schema._properties.IsNullOrEmpty())
+                {
+                    _insensitivePropertySchemas?.Clear();
+                }
+                else
+                {
+                    _insensitivePropertySchemas = new JSchemaDictionary(schema, new Dictionary<string, JSchema>(schema._properties, StringComparer.OrdinalIgnoreCase));
+                }
+
+                if (schema._dependencies.IsNullOrEmpty())
+                {
+                    _insensitiveDependencySchemas?.Clear();
+                }
+                else
+                {
+                    _insensitiveDependencySchemas = new JSchemaDependencyDictionary(schema, new Dictionary<string, object>(schema._dependencies, StringComparer.OrdinalIgnoreCase));
+                }
+
+                if (schema._dependentRequired.IsNullOrEmpty())
+                {
+                    _insensitiveDependentRequired?.Clear();
+                }
+                else
+                {
+                    _insensitiveDependentRequired = new Dictionary<string, IList<string>>(schema._dependentRequired, StringComparer.OrdinalIgnoreCase);
+                }
+            }
```

Comment:
This will be quite expensive.

Instead of copying these dictionaries with a new comparer, what about adding get methods for them to `ObjectScope`, e.g. `GetDependentRequired(string key)`.

If `PropertyNameCaseInsensitive` is false then use the dictionary indexer. If it is true, then foreach over the dictionary and check the key with `string.Equals(key, kvp.Key, StringComparison.OrdinalIgnoreCase)`. Note: try to avoid allocating `IEnumerable<T>` instances when doing this.

The collections aren't huge so the overhead of not using the indexer shouldn't be significant. And it should be cheaper than recreating and copying collections.

Reply:
Makes sense. Updated.